isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+		{
+			return write_bits((std::uint8_t const *)&data, sizeof (T) * 8);
+		}
+
+		template <typename T>
+		bool write(T const & data, std::size_t bits)
+		{
+			return write_bits((std::uint8_t const *)&data, bits);
+		}
+
+		template <>
+		bool write<bool>(bool const & data)
+		{
+			std::uint8_t bits = data ? 255 : 0;
+			return write_bits(&bits, 1);
+		}
+
+		template <>
+		bool write<char const *>(char const * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<char *>(char * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t const *>(std::uint8_t const * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		template <>
+		bool write<std::uint8_t *>(std::uint8_t * const & data)
+		{
+			// What should the default for including NULL be?
+			return write((std::uint8_t const *)data, false);
+		}
+
+		bool write(char const * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(char * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool write(std::uint8_t const * data, bool includeNull)
+		{
+			// Base case.  Write each byte separately so they don't get put in little-endian, which
+			// makes no sense for strings.
+			std::size_t revert = writeOffset;
+			while (*data)
+			{
+				if (!write_bits(data, 8))
+				{
+					writeOffset = revert;
+					return false;
+				}
+				++data;
+			}
+			if (includeNull)
+			{
+				std::uint8_t naught = 0;
+				if (!write_bits(&naught, 8))
+				{
+					writeOffset = revert;
+					return false;
+				}
+			}
+			return true;
+		}
+
+		bool write(std::uint8_t * data, bool includeNull)
+		{
+			return write((std::uint8_t const *)data, includeNull);
+		}
+
+		bool pad(std::size_t bits, bool value = true)
+		{
+			std::uint8_t data = value ? 255 : 0;
+			std::size_t byte = get_write_byte_offset();
+			std::size_t offset = get_write_bit_offset();
+			std::size_t remaining = 8 - offset;
+			std::uint8_t mask = 255 >> remaining;
+			writeOffset += bits;
+			buffer.resize((writeOffset + 7) / 8, 255);
+			for (;;)
+			{
+				buffer[byte] = (buffer[byte] & mask) | (data << offset);
+				if (remaining > bits)
+				{
+					// Weirdly we need a second masking here, to keep untouched bits as `1`.
+					buffer[byte] |= 255 << (bits + offset);
+					break;
+				}
+				else if (remaining == bits)
+				{
+					// Done.
+					break;
+				}
+				else
+				{
+					bits -= remaining;
+					offset = 0;
+					remaining = 8;
+					mask = 0;
+					++byte;
+				}
+			}
+			return true;
+		}
+
+		template <typename T>
+		bool read(T & data)
+		{
+			return read_bits((std::uint8_t *)&data, sizeof (T) * 8);
```

Comment:
Casting to `uint8_t *` from T is not endianness friendly, plus it's a C style cast which should not be used. This happens multiple places, and may require a fairly substantial refactor to fully resolve.